src/ledger/LedgerTxn.cpp

```diff
@@ -496,68 +550,189 @@ LedgerTxn::Impl::getAllOffers()
+    // inaccurate.
+    for (auto const& kv : mActive)
     {
-        auto const& key = iter->first;
-        auto const& entry = iter->second;
-        if (key.type() != OFFER)
+        if (kv.first.type() == OFFER)
         {
-            continue;
+            throw std::runtime_error(
+                "Active offer found during prepareGetBestOffer");
         }
+    }
 
-        if (!exclude.insert(key).second)
-        {
-            continue;
-        }
+    std::unique_ptr<GetBestOffersState> gbos = std::move(mGetBestOffersState);
 
-        if (!(entry && entry->data.offer().buying == buying &&
-              entry->data.offer().selling == selling))
-        {
-            continue;
-        }
+    // The GetBestOffersState will be reset in create, erase, and load (and some
+    // other similar functions) if it can no longer be guaranteed to be accurate
+    // so we can accept it as accurate if it meets these conditions.
+    if (!(gbos && gbos->mBuying == buying && gbos->mSelling == selling))
+    {
+        gbos = std::make_unique<GetBestOffersState>(buying, selling);
 
-        if ((bestOfferIter == end) ||
-            isBetterOffer(*entry, *bestOfferIter->second))
+        for (auto const& kv : mEntry)
         {
-            bestOfferIter = iter;
+            auto const& key = kv.first;
+            if (key.type() == OFFER)
+            {
+                auto const& entry = kv.second;
+                if (entry && entry->data.offer().buying == buying &&
+                    entry->data.offer().selling == selling)
+                {
+                    gbos->mOffers.emplace(*entry);
+                }
+            }
         }
     }
 
-    std::shared_ptr<LedgerEntry const> bestOffer;
-    if (bestOfferIter != end)
+    // Calling prepareGetBestOffer must always reset the stream to the beginning.
+    gbos->mIter = gbos->mOffers.cbegin();
+
+    // At this point mGetBestOffersState = nullptr but the pending value of
+    // mGetBestOffersState has already been computed. Therefore we can say the
+    // following:
+    // - If self or any parent throws an exception during prepareGetBestOffer
+    //   then mGetBestOffersState = nullptr
+    // - Otherwise mGetBestOffersState satisfies conditions 2a, 2b, 2c, and 2d
+    //   from the documentation of mGetBestOffersState
+    mParent.prepareGetBestOffer(buying, selling);
+
+    // Move-assignment of std::unique_ptr<...> never throws
+    mGetBestOffersState = std::move(gbos);
+}
+
+std::shared_ptr<LedgerEntry const>
+LedgerTxn::getBestOffer(Asset const& buying, Asset const& selling,
+                        LedgerKey const* exclude)
+{
+    return getImpl()->getBestOffer(buying, selling, exclude);
+}
+
+std::shared_ptr<LedgerEntry const>
+LedgerTxn::Impl::getBestOffer(Asset const& buying, Asset const& selling,
+                              LedgerKey const* exclude)
+{
+    auto const& gbos = mGetBestOffersState;
+    if (!gbos && (mChild || !mEntry.empty()))
```

Comment:
Isn't this just `if (!gbos)` ? I don't think it's valid to call `getBestOffer` without calling `prepareGetBestOffer` ?

Reply:
Actually, the most efficient way to use `prepareGetBestOffer` and `getBestOffer` is to:
1. Call `prepareGetBestOffer` on a `LedgerTxn`
2. Open a child `LedgerTxn`
3. Call `loadBestOffer` on the child (_without calling_ `prepareGetBestOffer` on the child)
4. Do things other than `loadBestOffer`
4. Commit the child
5. Repeat steps 2-4 as necessary

Indeed, this is what we do in `convertWithOffers`.

Why should we not call `prepareGetBestOffer` on the child? Because `prepareGetBestOffer` always recursively resets the streams of all parents to the beginning, so this will lead to `O(n^2)` iterations when making `O(n)` calls to `prepareGetBestOffer` followed by `getBestOffer`.

Why is it safe to not call `prepareGetBestOffer` when `!mChild && mEntry.empty()`? Because in this case there are no offers in the `LedgerTxn` so the `GetBestOfferState` won't contain anything. Note this argument only works if `prepareGetBestOffer` was already called for the parent. Technically I don't even think that the `!mChild` part is required.

Despite the fact that this works, it is neither beautiful nor intuitive. Any ideas for making this part of the interface cleaner would be much appreciated.

Reply:
Ah yes, mmm, this is annoying. At a minimum it would be nice to enforce that `prepareGetBestOffer` was called on one of the parents. Right now I don't see how we can do this though